dogechain-lab/dogechain
# [jsonrpc] Add debug_traceTransaction endpoint

# Description

This PR provides a new jsonRPC namespace `debug` and `debug_traceTransaction` endpoint.
Provides only basic feature, no configuration, let alone different types of js tracers.
Those features shall be done in other PRs.

# Changes include


# Checklist


## Testing


### Manual tests

* Set up a one-node `geth` private network, and a one-node `dogechain` private network.
* Deploy some test contracts to each network, code as below:
* Call the contract methods in each network.
* Call `debug_traceTransaction` of all transactions in each network.
  * Should get the almost same results in each network. Small differences should be in gas prices only.
  * The comparison includes: pc, memory, stack, storage, return value, and formatting.

```solidity
pragma solidity ^0.8.0;

contract test3 {
    uint x;
    uint y;

    function set(uint _x, uint _y) public {
        x = _x;
        y = _y;
    }

    function getAdd() public view returns(uint) {
        return x + y;
    }
}
```

```solidity
pragma solidity ^0.8.0;

contract RevertTest {
    function revert1() public {
       require(1 == 0, "I make a mistake");
    }

    function revert2() public {
        // don't tell anybody the reason
        require(1 == 0);
    }
}
```

```solidity
pragma solidity ^0.4.23;

contract Calltest {
    address public b;
    uint256 public n;
 
    function test() public returns (address a){
        a=address(this);
        b=a;
    }

    function increase(uint number) public {
        n += number;
    }
}

contract Compare {
    address public b;
    address public testaddress;

    constructor(address _addressOfCalltest) public {
        testaddress = _addressOfCalltest;
    }

    function withcall() public {
        testaddress.call(bytes4(keccak256("test()")));
        testaddress.call(bytes4(keccak256("increase(uint256)")), 10);
    }

    function withdelegatecall() public {
        testaddress.delegatecall(bytes4(keccak256("test()")));
        testaddress.call(bytes4(keccak256("increase(uint256)")), 100);
    }
}
```

# Additional comments

Associated issues: #50
